test/test_utils/ec2.py

```diff
@@ -452,6 +458,73 @@ def get_ec2_instance_tags(instance_id, region=DEFAULT_REGION, ec2_client=None):
     return {tag["Key"]: tag["Value"] for tag in response.get("Tags")}
 
 
+def get_last_line_of_s3_file(s3_location, local_filename="temp.txt"):
+    """
+    Extracts the last line of s3 file by copying it locally.
+    
+    :param s3_location: str, s3 uri
+    :param local_filename: str, location where s3 file is to be downloaded locally.
+    :return: str, The last line of the file
+    """
+    run(f"rm -rf {local_filename}", hide=True)
+    run(f"aws s3 cp {s3_location} {local_filename}", hide=True)
+    last_line_of_file = run(f"tail -n1 {local_filename}", hide=True).stdout.strip()
+    return last_line_of_file
+
+
+def execute_asynchronus_testing_using_s3_bucket(
+    connection,
+    execution_command,
+    connection_timeout,
+    s3_location,
+    required_log_ending,
+    loop_time=2.5 * 3600,
+    loop_sleep_time=5 * 60,
```

Comment:
Do you really need such flexibility? I'd suggest to remove it from function parameters. 
Imagine somebody will pass `loop_sleep_time=1`. The test will always fail because there will be no log update during 3 seconds.

And `loop_time` should be a bit less or equal to proper test job timeout parameter. Otherwise you'll have a hard time investigating tests timeout failures

Reply:
I have removed `loop_sleep_time` as it was too granular. The 'connection_timeout` and `loop_time` are different. `connection_timeout` is the generic fabric connection timeout and `loop_time` is the amount of time for which we would run the loop and is more flexible and can be changed based on tests.